Allow forced replacement of route table associated with subnet

YakDriver opened the pull request:
<!--- Information about referencing Github Issues: https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests --->
Fixes #73 

**UPDATE 7/19/19:** Rather than force replacement, allow the import of route table associations allows the associations to be updated. See [below](https://github.com/terraform-providers/terraform-provider-aws/pull/6999#issuecomment-513407131).

Changes proposed in this pull request:

* ~~`r/aws_route_table_association`: Add new argument `force_replace` allowing a route table to be associated with a subnet even if another route table is already associated.~~ Add corresponding docs & tests.
* `r/aws_route_table_association`: Implement import so that associations can be updated after they are imported.


Output from acceptance testing:

```console
$ make testacc TESTARGS='-run=TestAccAWSRouteTableAssociation_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccAWSRouteTableAssociation_ -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSRouteTableAssociation_basic
=== PAUSE TestAccAWSRouteTableAssociation_basic
=== RUN   TestAccAWSRouteTableAssociation_replace
=== PAUSE TestAccAWSRouteTableAssociation_replace
=== CONT  TestAccAWSRouteTableAssociation_basic
=== CONT  TestAccAWSRouteTableAssociation_replace
--- PASS: TestAccAWSRouteTableAssociation_replace (63.32s)
--- PASS: TestAccAWSRouteTableAssociation_basic (63.59s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	63.696s
```

YakDriver commented:
Here is a usage example. 

Assume that `subnet-04ff24b9b39f3822b` was already associated with a route table. Normally, attempting to associate another route table, you would receive an error like this:

```
* aws_route_table_association.new_assoc: Resource.AlreadyAssociated: the specified association for route table rtb-05c7d040bf39801b0 conflicts with an existing association
```

However, with this PR, you can set the `force_replace` argument to true, and the association with replace the prior association.

```hcl
resource "aws_route_table_association" "new_assoc" {
  route_table_id = "${aws_route_table.new_guy.id}"
  subnet_id      = "subnet-04ff24b9b39f3822b"
  force_replace  = true
}
```

```console
$ terraform apply
aws_route_table.new_guy: Creation complete after 0s (ID: rtb-0833a30c3ba11dc03)
aws_route_table_association.new_assoc: Creating...
  force_replace:  "" => "true"
  route_table_id: "" => "rtb-0833a30c3ba11dc03"
  subnet_id:      "" => "subnet-04ff24b9b39f3822b"
aws_route_table_association.new_assoc: Creation complete after 1s (ID: rtbassoc-08cbb18f6a3d7a975)

Apply complete! Resources: 2 added, 0 changed, 0 destroyed.
```

yves-vogl commented:
I'm hitting this issue too and would like to thank you for this patch.

janjur commented:
Please, merge.

aleksap commented:
Please merge! We are hitting a major blocker with this issue. 
Thanks!

ibehren1 commented:
Please merge -- need this as well.

pit commented:
Waiting for merge also!

My case is importing existing resources into terraform state. And it's not possible to import aws_route_table_association (https://github.com/terraform-providers/terraform-provider-aws/issues/561) and not possible to force old route table recreation.

lorengordon commented:
Anything we can do to promote this patch? Keep having to pop into the console to deal with this one, when people re-associate a route table improperly, since terraform fails to override this association. :(

aleksap commented:
Any good soul out there that could merge this? ^^

YakDriver commented:
@bflad As discussed, I've implemented import for route table associations and taken out the `force_replace` argument. I've updated the docs and acceptance tests to test imports and updates.

```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSRouteTableAssociation_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSRouteTableAssociation_ -timeout 120m
=== RUN   TestAccAWSRouteTableAssociation_basic
=== PAUSE TestAccAWSRouteTableAssociation_basic
=== RUN   TestAccAWSRouteTableAssociation_replace
=== PAUSE TestAccAWSRouteTableAssociation_replace
=== CONT  TestAccAWSRouteTableAssociation_basic
=== CONT  TestAccAWSRouteTableAssociation_replace
--- PASS: TestAccAWSRouteTableAssociation_basic (84.47s)
--- PASS: TestAccAWSRouteTableAssociation_replace (87.10s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	87.162s
```

YakDriver commented:
@aleksap @lorengordon @pit @ibehren1 @janjur @yves-vogl Please speak up if there are scenarios where import -> modify won't work and that you'd still need to force replace an association. My thinking is that having import capability (this PR) should cover most/all scenarios.

bflad commented on aws/resource_aws_route_table_association.go:
```diff
@@ -153,3 +157,49 @@ func resourceAwsRouteTableAssociationDelete(d *schema.ResourceData, meta interfa
+
+	var associationID string
+	for _, a := range rt.Associations {
+		if *a.SubnetId == subnetID {
```
To prevent potential panics, we prefer to use the [AWS Go SDK provided conversion functions](https://docs.aws.amazon.com/sdk-for-go/api/aws/), e.g. `aws.StringValue()` 👍 

```suggestion
		if aws.StringValue(a.SubnetId) == subnetID {
```

bflad commented on aws/resource_aws_route_table_association.go:
```diff
@@ -153,3 +157,49 @@ func resourceAwsRouteTableAssociationDelete(d *schema.ResourceData, meta interfa
+	var associationID string
+	for _, a := range rt.Associations {
+		if *a.SubnetId == subnetID {
+			associationID = *a.RouteTableAssociationId
```
Same here 👍 

```suggestion
			associationID = aws.StringValue(a.RouteTableAssociationId)
```

bflad commented on website/docs/r/route_table_association.html.markdown:
```diff
@@ -32,3 +32,17 @@ In addition to all arguments above, the following attributes are exported:
 
+## Import
+
+~> **NOTE:** Attempting to associate a route table with a subnet, where either
```
We should be covering this in the future with https://github.com/terraform-providers/terraform-provider-aws/issues/9223 😄

YakDriver replied:
You want the note removed?

YakDriver commented on aws/resource_aws_route_table_association_test.go:
```diff
@@ -38,6 +38,39 @@ func TestAccAWSRouteTableAssociation_basic(t *testing.T) {
+				ResourceName:      resourceName,
+				ImportState:       true,
+				ImportStateIdFunc: testAccAWSRouteTabAssocImportStateIdFunc(resourceName),
+				ImportStateVerify: true,
```
Sure

YakDriver commented on aws/resource_aws_route_table_association.go:
```diff
@@ -47,17 +50,18 @@ func resourceAwsRouteTableAssociationCreate(d *schema.ResourceData, meta interfa
 				}
 			}
 			return resource.NonRetryableError(err)
+		} else {
```
Definitely

YakDriver commented:
@bflad I made the changes you requested. Let me know if there is anything else you see.

```console
$ make testacc TESTARGS='-run=TestAccAWSRouteTableAssociation_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -parallel 20 -run=TestAccAWSRouteTableAssociation_ -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSRouteTableAssociation_basic
=== PAUSE TestAccAWSRouteTableAssociation_basic
=== RUN   TestAccAWSRouteTableAssociation_replace
=== PAUSE TestAccAWSRouteTableAssociation_replace
=== CONT  TestAccAWSRouteTableAssociation_basic
=== CONT  TestAccAWSRouteTableAssociation_replace
--- PASS: TestAccAWSRouteTableAssociation_basic (69.87s)
--- PASS: TestAccAWSRouteTableAssociation_replace (70.10s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	70.190s
```
